File: test_samples/java/vulnerable.java
package com.example;

import javax.servlet.http.*;
import javax.xml.parsers.*;
import org.xml.sax.*;
import com.thoughtworks.xstream.XStream;
import java.io.*;
import java.security.MessageDigest;
import java.sql.*;
import java.nio.file.*;

// ==================== JAVA-SEC-001: SQL Injection (CRITICAL) ====================

public class SqlHandler {
    public void badQuery(String userId) throws SQLException {
        Connection conn = DriverManager.getConnection(url, user, pass);
        Statement stmt = conn.createStatement();
        String sql = "SELECT * FROM users WHERE id = " + userId;
        ResultSet rs = stmt.executeQuery(sql);
        // Line: SQL injection with Statement
    }

    public void badQuery2(String name) throws SQLException {
        Connection conn = DriverManager.getConnection(url, user, pass);
        String sql = "SELECT * FROM products WHERE name = '" + name + "'";
        stmt.executeUpdate(sql);
        // Line: SQL concat
    }
}

// ==================== JAVA-SEC-002: Unsafe Deserialization (CRITICAL) ====================

class SerialHandler {
    public Object deserialize(byte[] data) throws Exception {
        ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(data));
        Object obj = ois.readObject();
        // Line: ObjectInputStream RCE
        return obj;
    }

    public void xmlDeserialize(String xml) {
        XMLDecoder decoder = new XMLDecoder(new ByteArrayInputStream(xml.getBytes()));
        Object obj = decoder.readObject();
        // Line: XMLDecoder unsafe
    }
}

// ==================== JAVA-SEC-003: XXE (CRITICAL) ====================

class XxeHandler {
    public void parseXml(String xml) throws Exception {
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        DocumentBuilder builder = factory.newDocumentBuilder();
        Document doc = builder.parse(new InputSource(new StringReader(xml)));
        // Line: XXE in DocumentBuilderFactory
    }

    public void parseSax(String xml) throws Exception {
        SAXParserFactory spf = SAXParserFactory.newInstance();
        SAXParser parser = spf.newSAXParser();
        // Line: XXE in SAXParserFactory
    }
}

// ==================== JAVA-SEC-004: Path Traversal (HIGH) ====================

class FileHandler {
    public String readFile(String filename) throws IOException {
        FileInputStream fis = new FileInputStream(filename + ".txt");
        // Line: FileInputStream path traversal
        return readStream(fis);
    }

    public void readPath(String userInput) throws IOException {
        Path p = Paths.get("/data/" + userInput);
        // Line: Paths.get traversal
        String content = new String(Files.readAllBytes(p));
    }
}

// ==================== JAVA-SEC-005: Command Injection (CRITICAL) ====================

class CmdHandler {
    public void runCommand(String userArg) throws Exception {
        Runtime rt = Runtime.getRuntime();
        Process p = rt.exec("ls " + userArg);
        // Line: Runtime.exec command injection
    }

    public void runProcessBuilder(String cmd) {
        ProcessBuilder pb = new ProcessBuilder(cmd + " -la");
        // Line: ProcessBuilder injection
    }
}

// ==================== JAVA-SEC-006: Hardcoded Secrets (HIGH) ====================

class Config {
    private static final String API_KEY = "sk-live-abc123xyz456secret789key12345";
    public static final String DB_PASSWORD = "MySecretPass123!";
    public static final String JWT_SECRET = "super_secret_jwt_key_do_not_use";
    private String awsKey = "AKIAIOSFODNN7EXAMPLE";
    private String authToken = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9";
    // Line: multiple hardcoded secrets
}

// ==================== JAVA-SEC-007: Weak Crypto (MEDIUM) ====================

class CryptoHandler {
    public String hashMd5(String input) throws Exception {
        MessageDigest md = MessageDigest.getInstance("MD5");
        byte[] digest = md.digest(input.getBytes());
        // Line: MD5 weak crypto
        return bytesToHex(digest);
    }

    public String hashSha1(String input) throws Exception {
        MessageDigest md = MessageDigest.getInstance("SHA1");
        byte[] digest = md.digest(input.getBytes());
        // Line: SHA1 deprecated
        return bytesToHex(digest);
    }

    public void useDes(byte[] key, byte[] data) throws Exception {
        javax.crypto.Cipher cipher = javax.crypto.Cipher.getInstance("DES/ECB/PKCS5Padding");
        // Line: DES weak
    }
}

// ==================== JAVA-SEC-008: LDAP Injection (MEDIUM) ====================

class LdapHandler {
    public void search(String username) throws Exception {
        DirContext ctx = new InitialDirContext(env);
        SearchControls controls = new SearchControls();
        String filter = "(&(objectClass=user)(cn=" + username + "))";
        ctx.search("ou=people", filter, controls);
        // Line: LDAP injection
    }
}

// ==================== Clean code (no issues) ====================

class SafeHandler {
    public void safeQuery(String userId) throws SQLException {
        Connection conn = DriverManager.getConnection(url, user, pass);
        PreparedStatement ps = conn.prepareStatement("SELECT * FROM users WHERE id = ?");
        ps.setString(1, userId);
        ResultSet rs = ps.executeQuery();
        // Parameterized query - safe
    }

    public void safeDeserialize(byte[] data) throws Exception {
        // Use JSON (Jackson/Gson) instead of Java serialization
    }

    public void safeXml(String xml) throws Exception {
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        factory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
        factory.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
        // XXE protected
    }
}

File: test_samples/csharp/vulnerable.cs
using System;
using System.IO;
using System.Web;
using System.Web.Mvc;
using System.Security.Cryptography;
using System.Data.SqlClient;

// ==================== CSHARP-SEC-001: Unsafe Deserialization (CRITICAL) ====================

class SerialHandler {
    public object Deserialize(byte[] data) {
        var formatter = new BinaryFormatter();
        var obj = formatter.Deserialize(new MemoryStream(data));
        // Line: BinaryFormatter RCE
        return obj;
    }

    public object DeserializeXml(string xml) {
        var serializer = new System.Runtime.Serialization.DataContractSerializer(typeof(object));
        // Verify input source before deserializing
    }
}

// ==================== CSHARP-SEC-002: SQL Injection (CRITICAL) ====================

class SqlHandler {
    public void Query(string userId) {
        string sql = "SELECT * FROM users WHERE id = " + userId;
        SqlCommand cmd = new SqlCommand(sql, conn);
        cmd.ExecuteReader();
        // Line: SQL injection SqlCommand
    }

    public void Query2(string name) {
        string query = "SELECT * FROM products WHERE name = '" + name + "'";
        SqlCommand cmd = new SqlCommand(query, conn);
        cmd.ExecuteNonQuery();
        // Line: SQL concat
    }
}

// ==================== CSHARP-SEC-003: Path Traversal (HIGH) ====================

class FileHandler {
    public void ReadFile(string filename) {
        string path = "/data/" + filename;
        var content = File.ReadAllText(path);
        // Line: File.ReadAllText path traversal
    }

    public void MapPath(string page) {
        string mapped = Server.MapPath("~/" + page);
        // Line: Server.MapPath traversal
    }
}

// ==================== CSHARP-SEC-004: LDAP Injection (MEDIUM) ====================

class LdapHandler {
    public void Search(string username) {
        using (var searcher = new System.DirectoryServices.DirectorySearcher()) {
            string filter = "(&(objectClass=user)(cn=" + username + "))";
            searcher.Filter = filter;
            // Line: LDAP injection
        }
    }
}

// ==================== CSHARP-SEC-005: Hardcoded Secrets (HIGH) ====================

class Config {
    private static readonly string API_KEY = "sk-live-abc123xyz456secret789key12345";
    public const string DB_PASSWORD = "MySecretPass123!";
    private string awsKey = "AKIAIOSFODNN7EXAMPLE";
    public string ConnectionString = "Server=localhost;Database=db;User=admin;Password=secret123";
    // Line: multiple hardcoded secrets
}

// ==================== CSHARP-SEC-006: Weak Crypto (MEDIUM) ====================

class CryptoHandler {
    public string HashMd5(string input) {
        using (var md5 = new MD5CryptoServiceProvider()) {
            // Line: MD5CryptoServiceProvider
            return BitConverter.ToString(md5.ComputeHash(System.Text.Encoding.UTF8.GetBytes(input)));
        }
    }

    public void UseDes(byte[] key, byte[] data) {
        var des = new DESCryptoServiceProvider();
        // Line: DESCryptoServiceProvider weak
    }
}

// ==================== CSHARP-SEC-007: XSS (HIGH) ====================

class XssHandler : Controller {
    public void Render(string content) {
        Response.Write("<div>" + content + "</div>");
        // Line: Response.Write XSS
    }

    public IHtmlString RenderRaw(string html) {
        return Html.Raw(html);
        // Line: Html.Raw XSS
    }
}

// ==================== Clean code (no issues) ====================

class SafeHandler {
    public void SafeQuery(string userId) {
        var cmd = new SqlCommand("SELECT * FROM users WHERE id = @id", conn);
        cmd.Parameters.AddWithValue("@id", userId);
        // Parameterized query - safe
    }

    public void SafeHash(string input) {
        using (var sha = new SHA256Managed()) {
            // SHA-256 - safe
        }
    }
}
